packages/@sanity/client/src/config.js

```diff
@@ -89,13 +76,60 @@ exports.initConfig = (config, prevConfig) => {
     const cdnHost = newConfig.isDefaultApi ? defaultCdnHost : host
 
     if (newConfig.useProjectHostname) {
-      newConfig.url = `${protocol}://${newConfig.projectId}.${host}/v1`
-      newConfig.cdnUrl = `${protocol}://${newConfig.projectId}.${cdnHost}/v1`
+      newConfig.url = `${protocol}://${newConfig.projectId}.${host}/v${newConfig.apiVersion}`
+      newConfig.cdnUrl = `${protocol}://${newConfig.projectId}.${cdnHost}/v${newConfig.apiVersion}`
     } else {
-      newConfig.url = `${newConfig.apiHost}/v1`
+      newConfig.url = `${newConfig.apiHost}/v${newConfig.apiVersion}`
       newConfig.cdnUrl = newConfig.url
     }
   }
 
   return newConfig
 }
+
+exports.validateApiVersion = function validateApiVersion(apiVersion, nowOffset = 0) {
+  if (apiVersion === '1' || apiVersion === 'X') {
+    return
+  }
+
+  const apiDate = new Date(apiVersion)
+  const apiVersionValid =
+    /^\d{4}-\d{2}-\d{2}$/.test(apiVersion) && apiDate instanceof Date && apiDate.getTime() > 0
+
+  if (!apiVersionValid) {
+    throw new Error('Invalid API version string, expected `1` or date in format `YYYY-MM-DD`')
+  }
+
+  const today = new Date(Date.now() + nowOffset)
+  const tomorrow = new Date(Date.now() + nowOffset)
+  tomorrow.setDate(today.getDate() + 1)
+
+  const todayUtc = getUtcDate(today)
+  const tomorrowUtc = getUtcDate(tomorrow)
+
+  if (apiVersion === tomorrowUtc) {
```

Comment:
I wonder whether we should have these sorts of checks in the client or not. We may want to change it (for whatever reason), and then the server- and client-side checks will no longer be in sync.

Reply:
Maybe the client should warn on dates in the future (because this is never a good idea in any case), but let the server-side API handle rejection of versions past tomorrow?

Reply:
This is the same check as we did in the PHP client (which you approved), but fair point.
It is kind of convoluted code that is prone to local clock drifts and whatnot, so if we all agree, I'd rather not include it.

Reply:
Yeah, didn't think of it until now. I'd say let's warn on future UTC dates (to help the poor Aussies who are bound to set their local date), but skip the rest - or do you think we should just remove it altogether?

Reply:
Not sure I get the distinction:
"Client should warn on dates in the future"
"Server-side API should reject on versions past tomorrow"

Do you mean that the client should warn if the date is beyond "tomorrow" UTC-wise, but leave the "tomorrow" check up to the server? If this is the case, I'd rather just remove the logic from the client at all, as date checks feel kind of brittle 🤷‍♂️